src/backend/distributed/metadata/dependency.c

```diff
@@ -0,0 +1,479 @@
+											HASH_ENTER, &found);
+
+	if (!found)
+	{
+		/* copy object address in */
+		*address = *collect;
+	}
+
+	/* add to list*/
+	collector->dependencyList = lappend(collector->dependencyList, address);
+}
+
+
+/*
+ * GetDependenciesForObject returns a list of ObjectAddesses to be created in order
+ * before the target object could safely be created on a worker. Some of the object might
+ * already be created on a worker. It should be created in an idempotent way.
+ */
+List *
+GetDependenciesForObject(const ObjectAddress *target)
+{
+	ObjectAddressCollector collector = { 0 };
+
+	InitObjectAddressCollector(&collector);
+
+	recurse_pg_depend(target,
+					  NULL,
+					  &FollowNewSupportedDependencies,
+					  &ApplyAddToDependencyList,
+					  &collector);
+
+	return collector.dependencyList;
+}
+
+
+/*
+ * IsObjectAddressCollected is a helper function that can check if an ObjectAddress is
+ * already in a (unsorted) list of ObjectAddresses
+ */
+static bool
+IsObjectAddressCollected(const ObjectAddress *findAddress,
+						 ObjectAddressCollector *collector)
+{
+	bool found = false;
+
+	/* add to set */
+	hash_search(collector->dependencySet, findAddress, HASH_FIND, &found);
+
+	return found;
+}
+
+
+/*
+ * SupportedDependencyByCitus returns whether citus has support to distribute the object
+ * addressed.
+ */
+static bool
+SupportedDependencyByCitus(const ObjectAddress *address)
+{
+	/*
+	 * looking at the type of a object to see if we know how to create the object on the
+	 * workers.
+	 */
+	switch (getObjectClass(address))
+	{
+		case OCLASS_SCHEMA:
+		{
+			return true;
+		}
+
+		default:
+		{
+			/* unsupported type */
+			return false;
+		}
+	}
+}
+
+
+/*
+ * IsObjectAddressOwnedByExtension returns whether or not the object is owned by an
+ * extension. It is assumed that an object having a dependency on an extension is created
+ * by that extension and therefore owned by that extension.
+ */
+static bool
+IsObjectAddressOwnedByExtension(const ObjectAddress *target)
+{
+	Relation depRel = NULL;
+	ScanKeyData key[2] = { 0 };
+	SysScanDesc depScan = NULL;
+	HeapTuple depTup = NULL;
+	bool result = false;
+
+	depRel = heap_open(DependRelationId, AccessShareLock);
+
+	/* scan pg_depend for classid = $1 AND objid = $2 using pg_depend_depender_index */
+	ScanKeyInit(&key[0], Anum_pg_depend_classid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(target->classId));
+	ScanKeyInit(&key[1], Anum_pg_depend_objid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(target->objectId));
+	depScan = systable_beginscan(depRel, DependDependerIndexId, true, NULL, 2, key);
+
+	while (HeapTupleIsValid(depTup = systable_getnext(depScan)))
+	{
+		Form_pg_depend pg_depend = (Form_pg_depend) GETSTRUCT(depTup);
+		if (pg_depend->deptype == DEPENDENCY_EXTENSION)
+		{
+			result = true;
+			break;
+		}
+	}
+
+	systable_endscan(depScan);
+	relation_close(depRel, AccessShareLock);
```

Comment:
should this be heap_close?

Reply:
possibly, I must have copied this somewhere from the postgres code, will replace.